File: Test Projects/Test All Projects/TestErrorHandlerBySelf/Form1.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Threading.Tasks;
using System.Windows.Forms;
using ErrorHandlerEngine.ModelObjecting;
//using ErrorLogAnalyzer;

namespace TestErrorHandlerBySelf
{
    public partial class Form1 : Form
    {
        private List<Action> Exps;

        public Form1()
        {
            InitializeComponent();

            Exps = new List<Action>
            {
                () => { int a = 0, b = 10, c = b/a; },
                () => { throw new ArithmeticException(); },
                () => { throw new Exception("TEst"); },
                () => { throw new InvalidExpressionException(); },
                () => { throw new ApplicationException(); },
                () => { throw new SystemException(); }
            };

            dataGridView1.CreateColumns(typeof(ProxyError));
        }

        private void btnQuit_Click(object sender, EventArgs e)
        {
            // End the program.
            Close();
        }

        private void dataGridView1_SelectionChanged(object sender, EventArgs e)
        {
            //int index = dataGridView1.CurrentRow != null ? dataGridView1.CurrentRow.Index : 0;
            //if (index < errors.Count && index >= 0)
            //    pictureBox1.Image = errors[index].Snapshot.Value;
        }

        private void btnTestHandledFirstExp_Click(object sender, EventArgs e)
        {
            var item = Exps[new Random().Next(0, Exps.Count - 1)];
            try
            {
                item();
            }
            catch (Exception ex)
            {
                ex.Data.Add("test1", "1");
                ex.Data.Add("test2", new Random().Next());
            }
        }

        private void btnTestUnHandledUIExp_Click(object sender, EventArgs e)
        {
            //Application.SetUnhandledExceptionMode(UnhandledExceptionMode.CatchException);
        }

        private void btnTestUnHandledThreadExp_Click(object sender, EventArgs e)
        {

            int a = 0, b = 10, c = b / a;
        }

        private void btnTestUnhandledTaskExp_Click(object sender, EventArgs e)
        {

            Task.Run(() =>
            { int a = 0, b = 10, c = b / a; });

            Task.Run(() =>
            { int a = 0, b = 8, c = b / a; });
        }

        private void btnRefreshGridView_Click(object sender, EventArgs e)
        {
            //ErrorLogReader.ReadAsync();
        }
    }
}

File: src/Test Projects/TestVBwinFormDotNet45/Form1.vb
Public Class Form1

    Private _exps As List(Of Action)

    Protected Overrides Sub OnLoad(ByVal e As EventArgs)
        MyBase.OnLoad(e)

        _exps = New List(Of Action)() From {
                Sub() Throw New Exception("Test"),
                Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New ArithmeticException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test"),
                 Sub() Throw New Exception("Test"),
                 Sub() Throw New InvalidExpressionException("Test"),
                 Sub() Throw New ApplicationException("Test"),
                 Sub() Throw New SystemException("Test")
            }
    End Sub

    Private Async Sub btnThrowHandledExps_Click(sender As Object, e As EventArgs) Handles btnThrowHandledExps.Click
        Await Task.Run(Sub() Parallel.[For](0, 8, Function(i)
                                                      Try
                                                          _exps(New Random().[Next](0, _exps.Count - 1))()
                                                      Catch
                                                      End Try
                                                  End Function))
    End Sub

    Private Sub btnThrowUnHandledExps_Click(sender As Object, e As EventArgs) Handles btnThrowUnHandledExps.Click
        _exps(New Random().[Next](0, _exps.Count - 1))()
    End Sub
End Class
